### PR TITLE
fix #104698 Make query element of uri survive conflict resolution massaging

### LINKED ISSUE #104698: FileSystemProvider whose uris use the query element is not properly handled by conflict resolution feature
I am developing an extension that implements a FileSystemProvider for which the query element of the uri is significant.

When an edited document is saved and VS Code detects that the target file is newer, VS Code correctly offers this notification:

image

But when the 'Compare' option is taken, the uri sent to the FSP has lost its query string.

The fault is here:

https://github.com/microsoft/vscode/blob/cc29a814476d9d9366e7687a16cff1af0641fbdf/src/vs/workbench/contrib/files/common/files.ts#L166-L172

`resourceToTextFile` sequesters the query element to stash the original scheme when setting a new scheme. Later `textFileToResource` reverses this process, but in doing so it has lost the query element of the original uri.

I will submit a PR.

### PR DESCRIPTION
**Pull request by gjsjohnmurray**
This PR fixes #104698. The fixed code originated with @bpasero

### CODE REVIEW

File: src/vs/workbench/contrib/files/common/files.ts [hunk 1]
```diff
@@ -164,11 +164,11 @@ export class TextFileContentProvider extends Disposable implements ITextModelCon
 	}
 
 	private static resourceToTextFile(scheme: string, resource: URI): URI {
-		return resource.with({ scheme, query: JSON.stringify({ scheme: resource.scheme }) });
+		return resource.with({ scheme, query: JSON.stringify({ scheme: resource.scheme, query: resource.query }) });
 	}
 
 	private static textFileToResource(resource: URI): URI {
-		return resource.with({ scheme: JSON.parse(resource.query)['scheme'], query: null });
+		return resource.with({ scheme: JSON.parse(resource.query)['scheme'], query: JSON.parse(resource.query)['query'] });
```
Comment (bpasero) [inline src/vs/workbench/contrib/files/common/files.ts:171]: Instead of calling `JSON.parse(resource.query)` twice, can this be extracted to a variable?

**Review by bpasero — CHANGES_REQUESTED** [review]

File: src/vs/workbench/contrib/files/common/files.ts (hunk 1 above)
Comment (gjsjohnmurray) [reply src/vs/workbench/contrib/files/common/files.ts]: I have pushed a change.

Comment (gjsjohnmurray) [reply src/vs/workbench/contrib/files/common/files.ts]: Actually, 3 changes, each simpler that the previous. Let me know if I have simplified to the point of obscurity and you'd like me to row back.

File: src/vs/workbench/contrib/files/common/files.ts [hunk 2]
```diff
@@ -168,7 +168,8 @@ export class TextFileContentProvider extends Disposable implements ITextModelCon
 	}
 
 	private static textFileToResource(resource: URI): URI {
-		return resource.with({ scheme: JSON.parse(resource.query)['scheme'], query: JSON.parse(resource.query)['query'] });
+		const { scheme, query } = JSON.parse(resource.query);
+		return resource.with({ scheme: scheme, query: query });
```
Comment (bpasero) [inline src/vs/workbench/contrib/files/common/files.ts]:
```suggestion
		return resource.with({ scheme, query });
```

**Review by bpasero — CHANGES_REQUESTED** [review]
Slight change suggested, otherwise looks good to me.

File: src/vs/workbench/contrib/files/common/files.ts (hunk 2 above)
Comment (gjsjohnmurray) [reply src/vs/workbench/contrib/files/common/files.ts]: I think it's now as you suggest.

**Comment by bpasero** [comment]
Thanks 👍

**Pull request merged**